Expected tuple struct or tuple variant, found struct variant [E0532]

#### I tried this code:
- for testing [`E0532`](https://doc.rust-lang.org/error_codes/E0532.html) error, Arm Does Not Match Expected Kind - expected tuple struct or tuple variant, found struct variant.
- Godbolt [link](https://godbolt.org/#z:OYLghAFBqd5QCxAYwPYBMCmBRdBLAF1QCcAaPECAMzwBtMA7AQwFtMQByARg9KtQYEAysib0QXACx8BBAKoBnTAAUAHpwAMvAFYTStJg1DEArgoKkl9ZATwDKjdAGFUtEywZ7HAGTwNMAHLuAEaYxBLSAA6oCoR2DC5uHnrRsbYCvv5BLKHhUpaY1ukMQgRMxASJ7p5cBUXxpeUEmYEhYRGWZRVVybXmTS3ZuREAlJaoJsTI7ByM7gDUjQSY8wCkAOwAQqsaAILzB4smyNOYWOikO/uHAGJMdGcQJgDMAExjVxsAIp97VAzzSLEPwEAD6AlBVHubmImAg/WWIDWrwAbEtMCM1lsrod5iwmARkAh5gAqBErDbbPa43EAelp8zCxBISMwqkimBsZ3mJgYhHm5lMNlIPL5BHmADdynhDOKSPM0Ax%2BoIRfxeeh5gQTJF6DiaQd6fqDlLgbK1iiNOiQCA7g90KsLXr9VabdDuatnj9nth5hAPtSjaC1p6PT6/U6sT89t8rhwxrROABWXieDhaUioTgAJTM4oUEymFNezx4pAImjjYwA1iBXomAHQaACcAA5Xk3E031ij26jagmOJIUxWM5xeAoQBoyxWxnBYDBEChUCxIg8yBQIGgV2uQMATsQFABaBTMSIKBCoCxYCV4aYANTwmAA7gB5Dmp0s0WjLA%2BUYIj4I/HKABPThS0A5hiGAl9gm0Tly24XgtzYQQXwYWhQLTXgsGCExgCcMRaAnRDSCwfEjHELDSLwWEbDwCVMGI9M2U5ExERIkFChHWg8GCYgQJcLARwIYEWDAuM%2BAMYAFAfZ830YcSZEEEQxHYfJ%2BEERQVHUKjdH7QxjFzfReInSAxlQSJimI3hUAY4hgSwUyoGYNgQDOOIBFIKU3BmV4NFeHgRjGKx4PiBwGGcVxqi8CLBjaPJSFSDyEii3pEpiZK4pydo%2BkKUKBEabpUpqOp8pKLpmj8VpsoS/oiqSEq6sqrJ4okYKC2mNr9CTYcqMzDh5lUFsUUPFFJHmABxJwnCzIRfSmma5vxcwwkxCBcEIeVVmLLgRl4BCtCC0ga0TKcByHUgxMkJt6xbRMey4LgUWeC07qbFFSFTdN%2BvHSdpyw2cFwgJAJgISI2PISgt1XehiACVgZlMcxkHmLh6xRFtG14dyiAcvQNOEURxHU2QtLUEc9NIJ9%2BMicT4x6z6R36l82PBuUqEG4bRvGhbZvm6a%2BeWn81pcbdYeREs9v%2Bw76cHXgxNO%2BspGedYWxbF7nibKQNGkL6bLHSw/oOytjpAZ5JHrRNzfWEsNCe1t1kkFtuo4Z5eu%2Bg3jcB%2BA50XaG10hzdlxh9piCkJsp2vW9MFk1930Ur8fwnCB/yoiCQMU9OoJguCbEU5DGAINCMJHHC8II2giMUsiDMo9N8Fo2wGKY7HVFY9jS04gd0x4viBIwGZ0xEvAxMQsYqCkmTHzjhSSIJlTiekAmyZ09NdFefQDJAJGLF7pzzMs%2BJrIzOyHMY%2BBgryuj7AgRwehKnwqqGHL0rSeJ75SDLiiy4ZcvqAqKofz/mVQqzVqq/06E0IBkCKg/xyu1SYnVdou2TIzPqnB5g7xRmja6jZfQbVxhLXa%2B0ZxjAQJgJgWBwh%2BlNlwDQZ1OAXQVlOPWo4OC/SnF7UgvtgYgFBmzQO/tYbw1chwIaI0xqTQFnNCAvMlpMBWsQKWOMSB4HQPjWQC81JL1JkocmulaxUxpnTFB7t9YcBZmDNi8xUAcywajesuCNC%2BlFiHYgEt3gkIBmQihVDKDVgkPQl2TCQCnTQR7dhhtOGkNNpIDQ9YXooi7M8ehKJ1hcFbO9F2btwnmK8TLF2rwzFsPySbOysR7CSCAA%3D).

```rust
enum State {
    Succeeded,
    Failed(u32),
}

fn print_on_failure(state: &State) {
    match *state {
        // error: expected unit struct, unit variant or constant, found tuple
        //        variant `State::Failed`
        State::Failed => (),
        _ => ()
    }
}

```

#### I expected to see this happen: 
Abort due to [`E0532`](https://doc.rust-lang.org/error_codes/E0532.html) error
```bash
error[E0532]: expected unit struct, unit variant or constant, found tuple variant `State::Failed`
  --> <source>:10:9
   |
3  |     Failed(u32),
   |     ----------- `State::Failed` defined here
...
10 |         State::Failed => (),
   |         ^^^^^^^^^^^^^ help: use the tuple variant pattern syntax instead: `State::Failed(_)`

error: aborting due to previous error
```
#### Instead, this happened:
Compiled with warning 
```bash
<source>:6:1: warning: function is never used: 'print_on_failure'
    6 | fn print_on_failure(state: &State) {
      | ^~
Compiler returned: 0
```
### Meta

- What version of Rust GCC were you using, git sha : ac43e5899ccb83b4d11ce1ae5e441f7b3734b9b2.
  - gccrs (Compiler-Explorer-Build-gcc-3ccf008b148455c4800f2a7c70be03e1dd708c59-binutils-2.40) 13.0.1 20230417 (experimental)

--- a/gcc/rust/typecheck/rust-hir-type-check-pattern.cc
+++ b/gcc/rust/typecheck/rust-hir-type-check-pattern.cc
@@ -44,6 +44,31 @@ void
 TypeCheckPattern::visit (HIR::PathInExpression &pattern)
 {
   infered = TypeCheckExpr::Resolve (&pattern);
+
+  /*
+   * We are compiling a PathInExpression, which can't be a Struct or Tuple
+   * pattern. We should check that the declaration we are referencing IS NOT a
+   * struct pattern or tuple with values.
+   */
+
+  rust_assert (infered->get_kind () == TyTy::TypeKind::ADT);
+  TyTy::ADTType *adt = static_cast<TyTy::ADTType *> (infered);
+
+  HirId variant_id = UNKNOWN_HIRID;
+  bool ok
+    = context->lookup_variant_definition (pattern.get_mappings ().get_hirid (),
+					  &variant_id);
+  rust_assert (ok);
+
+  TyTy::VariantDef *variant = nullptr;
+  ok = adt->lookup_variant_by_id (variant_id, &variant);
+
+  TyTy::VariantDef::VariantType vty = variant->get_variant_type ();
+
+  if (vty != TyTy::VariantDef::VariantType::NUM)
+    rust_error_at (
+      pattern.get_final_segment ().get_locus (), ErrorCode::E0532,
+      "expected unit struct, unit variant or constant, found tuple variant");
 }
 
 void

--- a/gcc/testsuite/rust/compile/issue-2324-1.rs
+++ b/gcc/testsuite/rust/compile/issue-2324-1.rs
@@ -1,0 +1,19 @@
+enum State {
+    Succeeded,
+    Failed(u32),
+}
+
+fn print_on_failure(state: &State) {
+    match *state {
+        State::Succeeded => (),
+        State::Failed => (), // { dg-error "expected unit struct, unit variant or constant, found tuple variant" }
+        _ => ()
+    }
+}
+
+fn main() {
+    let b = State::Failed(1);
+
+    print_on_failure(&b);
+
+}

--- a/gcc/testsuite/rust/compile/issue-2324-2.rs
+++ b/gcc/testsuite/rust/compile/issue-2324-2.rs
@@ -1,0 +1,19 @@
+enum State {
+    Succeeded,
+    Failed { x: u32 },
+}
+
+fn print_on_failure(state: &State) {
+    match *state {
+        State::Succeeded => (),
+        State::Failed => (), // { dg-error "expected unit struct, unit variant or constant, found tuple variant" }
+        _ => ()
+    }
+}
+
+fn main() {
+    let b = State::Failed{x: 1};
+
+    print_on_failure(&b);
+
+}

